File: contracts/agent_escrow.py
# { "Depends": "py-genlayer:1jb45aa8ynh2a9c9xn3b7qqh8sm5q93hwfp7jqmwsfhh8jpz09h6" }

"""AgentEscrow - a GenLayer Intelligent Contract for the Agent Tank hackathon.

One agent (the requester) escrows payment for another named agent (the
worker) to fetch and report a specific, checkable fact from a specific,
neutral third-party source. GenLayer validators independently re-fetch
the same source and judge whether the worker's report matches it before
releasing payment - the worker never gets paid on the requester's say-so,
and the requester can't refuse to pay for a correct report either.

Security model (why this shape, not a simpler one):
- The source must come from a hardcoded allowlist of neutral third
  parties, not a URL either side picks freely. Letting the worker name
  the source after doing the work would let them just point at a page
  they control; letting the requester name an arbitrary URL has the
  same problem in the other direction (they could tamper with it before
  adjudication to dodge paying). Only a source neither party controls
  makes the comparison meaningful.
- There is no time-based deadline (GenVM doesn't expose a deterministic
  clock to contracts). Instead the requester can cancel and reclaim
  their escrow at any time - but only while nothing has been submitted
  yet. Once a worker submits a report, cancellation is blocked; the task
  must go through adjudicate() and resolve to accepted or rejected,
  which auto-refunds the requester anyway if the report was wrong. This
  means a good-faith requester is never stuck, and a submitted report
  can't be silently discarded to dodge a payout.
- No appeals: unlike a subjective judgment call (e.g. "was this wallet
  really compromised?"), matching a reported value against a live-fetched
  source is close to deterministic, so a bad first verdict is unlikely
  enough that appeal complexity isn't worth it for this contract.
"""

from dataclasses import dataclass
from genlayer import *

# Neutral, third-party sources only - neither the requester nor the worker
# can be the operator of any of these, which is what makes the comparison
# meaningful. A source either side controlled could be tampered with to
# force whatever verdict benefits them.
ALLOWED_SOURCE_DOMAINS = [
    "api.coingecko.com",
    "api.coinbase.com",
    "etherscan.io",
]


def _is_allowed_source(source_url: str) -> bool:
    url = source_url.lower()
    return any(domain in url for domain in ALLOWED_SOURCE_DOMAINS)


@allow_storage
@dataclass
class Task:
    id: str
    requester: Address
    worker: Address
    source_url: str
    fact_description: str
    amount: u256
    status: str  # "open" | "submitted" | "accepted" | "rejected" | "cancelled"
    reported_value: str
    verdict_reasoning: str


class AgentEscrow(gl.Contract):
    """Escrows payment for a verified-data-delivery task between two
    agents, releasing it only once GenLayer validators confirm the
    worker's reported value matches an independently-fetched, neutral
    source."""

    tasks: TreeMap[str, Task]
    requester_tasks: TreeMap[Address, DynArray[str]]
    worker_tasks: TreeMap[Address, DynArray[str]]
    next_task_id: u256

    def __init__(self):
        self.next_task_id = 0

    @gl.public.write.payable
    def create_task(self, worker: str, source_url: str, fact_description: str) -> str:
        """Escrows this transaction's value and creates a task for
        `worker` to report `fact_description` as found at `source_url`.
        `source_url` must be on the neutral-source allowlist."""
        amount = gl.message.value
        if amount <= 0:
            raise gl.vm.UserError("Must escrow a positive amount to create a task")

        if not _is_allowed_source(source_url):
            raise gl.vm.UserError(
                "source_url must be one of the approved neutral data sources"
            )

        requester = gl.message.sender_address
        worker_address = Address(worker)

        task_id = str(self.next_task_id)
        self.next_task_id += 1

        task = Task(
            id=task_id,
            requester=requester,
            worker=worker_address,
            source_url=source_url,
            fact_description=fact_description,
            amount=amount,
            status="open",
            reported_value="",
            verdict_reasoning="",
        )
        self.tasks[task_id] = task
        self.requester_tasks.get_or_insert_default(requester).append(task_id)
        self.worker_tasks.get_or_insert_default(worker_address).append(task_id)
        return task_id

    @gl.public.write
    def cancel_task(self, task_id: str) -> None:
        """Refunds the requester in full - but only while the task is
        still open. Once a worker has submitted a report, the task must
        be resolved via adjudicate() instead, so a submitted report can
        never be silently discarded."""
        if task_id not in self.tasks:
            raise gl.vm.UserError("Task not found")

        task = self.tasks[task_id]
        if gl.message.sender_address != task.requester:
            raise gl.vm.UserError("Only the requester can cancel this task")
        if task.status != "open":
            raise gl.vm.UserError("Task already has a submitted report")

        task.status = "cancelled"
        gl.get_contract_at(task.requester).emit_transfer(value=task.amount)

    @gl.public.write
    def submit_deliverable(self, task_id: str, reported_value: str) -> None:
        """Only the named worker can submit, and only once, while the
        task is still open."""
        if task_id not in self.tasks:
            raise gl.vm.UserError("Task not found")

        task = self.tasks[task_id]
        if gl.message.sender_address != task.worker:
            raise gl.vm.UserError("Only the named worker can submit a deliverable")
        if task.status != "open":
            raise gl.vm.UserError("A deliverable was already submitted for this task")

        task.reported_value = reported_value
        task.status = "submitted"

    def _judge(self, source_url: str, fact_description: str, reported_value: str) -> dict:
        def leader_fn() -> dict:
            source_data = gl.nondet.web.render(source_url, mode="text")

            prompt = f"""
You are adjudicating a data-delivery task on AgentEscrow, a GenLayer Intelligent
Contract that pays one AI agent to fetch a fact for another.

Fact requested: {fact_description}
Authoritative source (fetched live, independently, by you - not provided or
editable by either party): {source_url}
\"\"\"
{source_data}
\"\"\"

The worker agent reported this value:
{reported_value}

Decide whether the reported value accurately reflects the fact as shown in the
authoritative source above. For a value that can fluctuate quickly (e.g. a live
price), allow for a small, reasonable margin to account for normal movement
between when the worker checked and when you fetched the source - but treat a
value that is clearly wrong, stale, fabricated, or contradicted by the source as
a reason to reject.

Respond in JSON:
{{
    "verdict": str,  // "accept" or "reject"
    "confidence": int,  // 0-100
    "reasoning": str  // one or two sentences
}}
It is mandatory that you respond only using the JSON format above,
nothing else. Don't include any other words or characters,
your output must be only JSON without any formatting prefix or suffix.
This result should be perfectly parsable by a JSON parser without errors.
"""
            return gl.nondet.exec_prompt(prompt, response_format="json")

        def validator_fn(leaders_res) -> bool:
            if not isinstance(leaders_res, gl.vm.Return):
                return False
            my_result = leader_fn()
            return my_result["verdict"] == leaders_res.calldata["verdict"]

        return gl.vm.run_nondet_unsafe(leader_fn, validator_fn)

    @gl.public.write
    def adjudicate(self, task_id: str) -> None:
        """Callable by anyone once a report has been submitted - not just
        the requester or worker - so neither side can stall the other's
        payout by refusing to trigger adjudication."""
        if task_id not in self.tasks:
            raise gl.vm.UserError("Task not found")

        task = self.tasks[task_id]
        if task.status != "submitted":
            raise gl.vm.UserError("Task has no pending report to adjudicate")

        verdict = self._judge(task.source_url, task.fact_description, task.reported_value)
        verdict_value = str(verdict.get("verdict", "")).lower()
        task.verdict_reasoning = str(verdict.get("reasoning", ""))

        if verdict_value == "accept":
            task.status = "accepted"
            gl.get_contract_at(task.worker).emit_transfer(value=task.amount)
        else:
            task.status = "rejected"
            gl.get_contract_at(task.requester).emit_transfer(value=task.amount)

    @gl.public.view
    def get_task(self, task_id: str) -> Task:
        if task_id not in self.tasks:
            raise gl.vm.UserError("Task not found")
        return self.tasks[task_id]

    @gl.public.view
    def get_tasks_by_requester(self, requester: str) -> list:
        addr = Address(requester)
        if addr not in self.requester_tasks:
            return []
        return [self.tasks[task_id] for task_id in self.requester_tasks[addr]]

    @gl.public.view
    def get_tasks_by_worker(self, worker: str) -> list:
        addr = Address(worker)
        if addr not in self.worker_tasks:
            return []
        return [self.tasks[task_id] for task_id in self.worker_tasks[addr]]

    @gl.public.view
    def get_all_tasks(self) -> dict:
        return {task_id: task for task_id, task in self.tasks.items()}

File: frontend/lib/contracts/AgentEscrow.ts
import { createClient } from "genlayer-js";
import { getGenLayerChain } from "../genlayer/chains";
import type { Task } from "./types";
import {
  estimateWriteFeePreset,
  feePresetToTransactionFees,
  type FeePresetEstimate,
  type FeePresetLevel,
} from "../genlayer/fees";

/**
 * genlayer-js decodes Python dataclasses (and dicts/TreeMaps) as JS Map
 * instances, keyed by field/dict-key name. This flattens one level of that
 * into a plain object, and normalizes Address-typed fields to a
 * checksummed hex string.
 */
function decodeTask(raw: any): Task {
  const entries = raw instanceof Map ? Array.from(raw.entries()) : Object.entries(raw ?? {});
  const obj: Record<string, any> = {};
  for (const [key, value] of entries) {
    obj[key] = value && typeof value === "object" && "as_hex" in value ? value.as_hex : value;
  }
  return {
    id: String(obj.id ?? ""),
    requester: String(obj.requester ?? ""),
    worker: String(obj.worker ?? ""),
    source_url: String(obj.source_url ?? ""),
    fact_description: String(obj.fact_description ?? ""),
    amount: String(obj.amount ?? "0"),
    status: (obj.status ?? "open") as Task["status"],
    reported_value: String(obj.reported_value ?? ""),
    verdict_reasoning: String(obj.verdict_reasoning ?? ""),
  };
}

/**
 * AgentEscrow contract class for interacting with the GenLayer
 * agent-to-agent verified-data-delivery escrow contract.
 */
class AgentEscrow {
  private contractAddress: `0x${string}`;
  private client: any;
  private rpcUrl?: string;

  constructor(contractAddress: string, address?: string | null, rpcUrl?: string) {
    this.contractAddress = contractAddress as `0x${string}`;
    this.rpcUrl = rpcUrl;

    const config: any = { chain: getGenLayerChain() };
    if (address) config.account = address as `0x${string}`;
    if (rpcUrl) config.endpoint = rpcUrl;

    this.client = createClient(config);
  }

  updateAccount(address: string): void {
    const config: any = { chain: getGenLayerChain(), account: address as `0x${string}` };
    if (this.rpcUrl) config.endpoint = this.rpcUrl;
    this.client = createClient(config);
  }

  async estimateCreateTaskFees(
    worker: string,
    sourceUrl: string,
    factDescription: string,
    level: FeePresetLevel = "standard"
  ): Promise<FeePresetEstimate | undefined> {
    return estimateWriteFeePreset(
      this.client,
      { address: this.contractAddress, functionName: "create_task", args: [worker, sourceUrl, factDescription] },
      level,
    );
  }

  async estimateSubmitDeliverableFees(
    taskId: string,
    reportedValue: string,
    level: FeePresetLevel = "standard"
  ): Promise<FeePresetEstimate | undefined> {
    return estimateWriteFeePreset(
      this.client,
      { address: this.contractAddress, functionName: "submit_deliverable", args: [taskId, reportedValue] },
      level,
    );
  }

  async estimateAdjudicateFees(
    taskId: string,
    level: FeePresetLevel = "standard"
  ): Promise<FeePresetEstimate | undefined> {
    return estimateWriteFeePreset(
      this.client,
      { address: this.contractAddress, functionName: "adjudicate", args: [taskId] },
      level,
    );
  }

  async getAllTasks(): Promise<Task[]> {
    try {
      const result: any = await this.client.readContract({
        address: this.contractAddress, functionName: "get_all_tasks", args: [],
      });
      if (result instanceof Map) return Array.from(result.values()).map(decodeTask);
      if (result && typeof result === "object") return Object.values(result).map(decodeTask);
      return [];
    } catch (error) {
      console.error("Error fetching tasks:", error);
      throw new Error("Failed to fetch tasks from contract");
    }
  }

  async getTask(taskId: string): Promise<Task | null> {
    try {
      const result = await this.client.readContract({
        address: this.contractAddress, functionName: "get_task", args: [taskId],
      });
      return decodeTask(result);
    } catch (error) {
      console.error("Error fetching task:", error);
      return null;
    }
  }

  async getTasksByRequester(address: string | null): Promise<Task[]> {
    if (!address) return [];
    try {
      const result: any = await this.client.readContract({
        address: this.contractAddress, functionName: "get_tasks_by_requester", args: [address],
      });
      return Array.isArray(result) ? result.map(decodeTask) : [];
    } catch (error) {
      console.error("Error fetching tasks by requester:", error);
      return [];
    }
  }

  async getTasksByWorker(address: string | null): Promise<Task[]> {
    if (!address) return [];
    try {
      const result: any = await this.client.readContract({
        address: this.contractAddress, functionName: "get_tasks_by_worker", args: [address],
      });
      return Array.isArray(result) ? result.map(decodeTask) : [];
    } catch (error) {
      console.error("Error fetching tasks by worker:", error);
      return [];
    }
  }

  /**
   * Creates a new task, escrowing `amountWei` as payment for `worker` to
   * report `factDescription` as found at `sourceUrl`.
   */
  async createTask(
    worker: string,
    sourceUrl: string,
    factDescription: string,
    amountWei: bigint,
    feePreset?: FeePresetEstimate,
    onSubmitted?: (txHash: string) => void
  ): Promise<string> {
    const fees = feePresetToTransactionFees(feePreset);
    let txHash: string;
    try {
      txHash = await this.client.writeContract({
        address: this.contractAddress,
        functionName: "create_task",
        args: [worker, sourceUrl, factDescription],
        value: amountWei,
        ...(fees ? { fees } : {}),
      });
    } catch (error) {
      console.error("Error creating task:", error);
      throw new Error("Failed to submit the create-task transaction. Please try again.");
    }

    onSubmitted?.(txHash);

    try {
      await this.client.waitForTransactionReceipt({ hash: txHash, status: "ACCEPTED" as any, retries: 24, interval: 5000 });
      return txHash;
    } catch (error) {
      console.error("Error confirming create-task transaction:", error);
      throw new Error(
        `Transaction ${txHash} was submitted but confirmation timed out. It may still complete - check the explorer.`
      );
    }
  }

  async cancelTask(taskId: string, onSubmitted?: (txHash: string) => void): Promise<string> {
    let txHash: string;
    try {
      txHash = await this.client.writeContract({
        address: this.contractAddress, functionName: "cancel_task", args: [taskId], value: BigInt(0),
      });
    } catch (error) {
      console.error("Error cancelling task:", error);
      throw new Error("Failed to submit the cancel transaction. Please try again.");
    }

    onSubmitted?.(txHash);

    try {
      await this.client.waitForTransactionReceipt({ hash: txHash, status: "ACCEPTED" as any, retries: 24, interval: 5000 });
      return txHash;
    } catch (error) {
      console.error("Error confirming cancel transaction:", error);
      throw new Error(
        `Transaction ${txHash} was submitted but confirmation timed out. It may still complete - check the explorer.`
      );
    }
  }

  async submitDeliverable(
    taskId: string,
    reportedValue: string,
    feePreset?: FeePresetEstimate,
    onSubmitted?: (txHash: string) => void
  ): Promise<string> {
    const fees = feePresetToTransactionFees(feePreset);
    let txHash: string;
    try {
      txHash = await this.client.writeContract({
        address: this.contractAddress,
        functionName: "submit_deliverable",
        args: [taskId, reportedValue],
        value: BigInt(0),
        ...(fees ? { fees } : {}),
      });
    } catch (error) {
      console.error("Error submitting deliverable:", error);
      throw new Error("Failed to submit the deliverable transaction. Please try again.");
    }

    onSubmitted?.(txHash);

    try {
      await this.client.waitForTransactionReceipt({ hash: txHash, status: "ACCEPTED" as any, retries: 24, interval: 5000 });
      return txHash;
    } catch (error) {
      console.error("Error confirming deliverable transaction:", error);
      throw new Error(
        `Transaction ${txHash} was submitted but confirmation timed out. It may still complete - check the explorer.`
      );
    }
  }

  async adjudicate(taskId: string, onSubmitted?: (txHash: string) => void): Promise<string> {
    const feePreset = await this.estimateAdjudicateFees(taskId);
    const fees = feePresetToTransactionFees(feePreset);
    let txHash: string;
    try {
      txHash = await this.client.writeContract({
        address: this.contractAddress, functionName: "adjudicate", args: [taskId], value: BigInt(0),
        ...(fees ? { fees } : {}),
      });
    } catch (error) {
      console.error("Error adjudicating task:", error);
      throw new Error("Failed to submit the adjudication transaction. Please try again.");
    }

    onSubmitted?.(txHash);

    try {
      await this.client.waitForTransactionReceipt({ hash: txHash, status: "ACCEPTED" as any, retries: 24, interval: 5000 });
      return txHash;
    } catch (error) {
      console.error("Error confirming adjudication transaction:", error);
      throw new Error(
        `Transaction ${txHash} was submitted but confirmation timed out. It may still complete - check the explorer.`
      );
    }
  }
}

export default AgentEscrow;
